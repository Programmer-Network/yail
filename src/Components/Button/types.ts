import { ReactNode } from "react";

import { IconName } from "Components/Icons/types";

export enum ButtonVariantEnum {
  PRIMARY = "primary",
  SECONDARY = "secondary"
}

export interface IButtonProps {
  onClick?: (e: React.MouseEvent) => void;
  children?: ReactNode;
  className?: string;
  disabled?: boolean;
  type?: "button" | "submit" | "reset";
  isLoading?: boolean;
  outlined?: boolean;
  icon?: {
    iconName?: IconName;
    iconClassName?: string;
    iconPosition?: "left" | "right";
  };
  variant?: ButtonVariantEnum;
}
