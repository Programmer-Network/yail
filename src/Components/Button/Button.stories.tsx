import type { Meta, StoryObj } from "@storybook/react";

import Button from ".";
import { ButtonVariantEnum } from "./types";

const meta = {
  title: "Core/Button",
  component: Button,
  tags: ["autodocs"],
  argTypes: {}
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Filled: Story = {
  args: {
    children: "Button"
  }
};

export const FilledDisabled: Story = {
  args: {
    children: "Button",
    disabled: true
  }
};

export const FilledSpining: Story = {
  args: {
    children: "Button",
    isLoading: true
  }
};

export const Outlined: Story = {
  args: {
    children: "Button",
    outlined: true
  }
};

export const OutlinedDisabled: Story = {
  args: {
    children: "Button",
    outlined: true,
    disabled: true
  }
};

export const OutlinedSpinning: Story = {
  args: {
    children: "Button",
    outlined: true,
    isLoading: true
  }
};

export const WithLeftIcon: Story = {
  args: {
    children: "Button",
    outlined: true,
    icon: {
      iconName: "IconBomb",
      iconPosition: "left"
    }
  }
};

export const WithRightIcon: Story = {
  args: {
    children: "Button",
    icon: {
      iconName: "IconBomb",
      iconPosition: "right"
    }
  }
};

export const WithIconClassName: Story = {
  args: {
    children: "Button",
    icon: {
      iconName: "IconBrain",
      iconPosition: "right",
      iconClassName: "text-violet-800 yl:w-5"
    }
  }
};

export const LoadingWithIcon: Story = {
  args: {
    children: "Button",
    outlined: true,
    isLoading: true,
    className: "button--green",
    icon: {
      iconName: "IconBomb",
      iconPosition: "left"
    }
  }
};

export const Red: Story = {
  args: {
    children: "Button",
    outlined: true,
    isLoading: true,
    className: "button--red"
  }
};

export const Gray: Story = {
  args: {
    children: "Button",
    outlined: true,
    isLoading: true,
    className: "button--gray"
  }
};

export const Green: Story = {
  args: {
    children: "Button",
    outlined: true,
    isLoading: true,
    className: "button--green"
  }
};

export const Secondary: Story = {
  args: {
    children: "Cancel",
    variant: ButtonVariantEnum.SECONDARY
  }
};

export const SecondaryDisabled: Story = {
  args: {
    children: "Cancel",
    variant: ButtonVariantEnum.SECONDARY,
    disabled: true
  }
};

export const SecondaryWithIcon: Story = {
  args: {
    children: "Cancel",
    variant: ButtonVariantEnum.SECONDARY,
    icon: {
      iconName: "IconAddCircle",
      iconPosition: "left"
    }
  }
};

export const SecondaryLoading: Story = {
  args: {
    children: "Cancel",
    variant: ButtonVariantEnum.SECONDARY,
    isLoading: true
  }
};

export const Primary: Story = {
  args: {
    children: "Submit",
    variant: ButtonVariantEnum.PRIMARY
  }
};
