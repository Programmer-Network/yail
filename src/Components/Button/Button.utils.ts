import classNames from "classnames";

import { ButtonVariantEnum, IGetButtonClassesArgs } from "./types";

export const getStyles = (isLoading: boolean) => {
  return {
    base: "yl:select-none yl:px-3 yl:py-2 yl:font-semibold yl:tracking-tight yl:rounded-md yl:border-2 yl:transition-all yl:duration-200 yl:transform yl:hover:scale-105",
    disabled: "yl:cursor-not-allowed yl:opacity-70",
    enabled: classNames({
      "yl:cursor-pointer": !isLoading
    }),
    primary: {
      default: "yl:border-primary yl:bg-primary yl:text-background",
      outlined:
        "yl:border-primary yl:bg-transparent yl:text-primary yl:fill-primary",
      hover: classNames({
        "yl:hover:bg-transparent yl:hover:text-primary": !isLoading,
        "yl:hover:fill-background": isLoading
      }),
      hoverOutlined: classNames({
        "yl:hover:bg-primary yl:hover:text-background yl:hover:fill-background":
          !isLoading
      })
    },
    error: {
      default: "yl:border-error yl:bg-error yl:text-background yl:fill-bg",
      outlined: "yl:border-error yl:bg-transparent yl:text-error yl:fill-error",
      hover: classNames({
        "yl:hover:bg-transparent yl:hover:text-error yl:hover:fill-error":
          !isLoading
      }),
      hoverOutlined: classNames({
        "yl:hover:bg-error yl:hover:text-background yl:hover:fill-background":
          !isLoading
      })
    },
    secondary: {
      default: "yl:border-text/20 yl:text-text/20 yl:fill-text/20",
      hover: classNames({
        "yl:hover:border-text/30 yl:hover:text-text/30 yl:hover:fill-text/30 yl:hover:bg-background/20":
          !isLoading
      })
    }
  };
};

export const getVariantClasses = (
  variant: ButtonVariantEnum,
  disabled: boolean,
  outlined: boolean,
  isLoading: boolean
) => {
  const styles = getStyles(isLoading);

  if (variant === ButtonVariantEnum.PRIMARY) {
    return outlined
      ? `${styles.primary.outlined} ${!disabled && styles.primary.hoverOutlined}`
      : `${styles.primary.default} ${!disabled && styles.primary.hover}`;
  } else if (variant === ButtonVariantEnum.ERROR) {
    return outlined
      ? `${styles.error.outlined} ${!disabled && styles.error.hoverOutlined}`
      : `${styles.error.default} ${!disabled && styles.error.hover}`;
  } else if (variant === ButtonVariantEnum.SECONDARY) {
    return `${styles.secondary.default} ${!disabled && styles.secondary.hover}`;
  }

  return "";
};

export const getButtonClasses = (args: IGetButtonClassesArgs) => {
  const { variant, disabled, outlined, isLoading, className } = args;

  const styles = getStyles(isLoading);

  return classNames(
    styles.base,
    disabled ? styles.disabled : styles.enabled,
    getVariantClasses(variant, disabled, outlined, isLoading),
    className
  );
};
