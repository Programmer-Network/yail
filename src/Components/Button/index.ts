export { default } from "./Button";
export * from "./types";
