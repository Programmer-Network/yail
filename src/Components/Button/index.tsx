import classNames from "classnames";

import Icon from "Components/Icon";
import Spinner from "Components/Spinner";

import "./button-style.css";
import { ButtonVariantEnum, IButtonProps } from "./types";

const Button: React.FC<IButtonProps> = (
  {
    onClick,
    children,
    className,
    disabled = false,
    type = "button",
    icon,
    isLoading = false,
    outlined = false,
    variant = ButtonVariantEnum.PRIMARY
  } = {
    onClick: () => null,
    children: null,
    className: "",
    disabled: false,
    outlined: false,
    isLoading: false,
    variant: ButtonVariantEnum.PRIMARY
  }
) => {
  let cls =
    "yl:select-none yl:px-3 yl:py-2 yl:font-semibold yl:tracking-tight yl:rounded-md ";

  if (variant === ButtonVariantEnum.PRIMARY) {
    cls += "yl:border-2 yl:border-primary ";
  }

  if (disabled) {
    cls += "yl:cursor-not-allowed yl:opacity-70 ";
    if (variant === ButtonVariantEnum.SECONDARY) {
      cls += "yl:text-text ";
    } else if (outlined) {
      cls += "yl:text-primary ";
    } else {
      cls += "yl:text-background yl:bg-primary ";
    }
  } else {
    if (variant === ButtonVariantEnum.SECONDARY) {
      cls += "yl:text-text yl:hover:opacity-80 ";
    } else if (outlined) {
      cls +=
        "yl:bg-transparent yl:text-primary yl:fill-primary yl:hover:bg-primary yl:hover:fill-background yl:hover:text-background";
    } else {
      cls +=
        "yl:bg-primary yl:hover:text-primary yl:hover:bg-transparent yl:hover:text-primary yl:hover:fill-primary yl:text-background";
    }
  }

  return (
    <button
      disabled={disabled}
      onClick={onClick}
      type={type}
      className={classNames(cls, className)}
    >
      <div className='yl:relative yl:flex yl:items-center yl:justify-center'>
        <span
          className={classNames({
            "yl:flex yl:gap-1 yl:items-center": icon?.iconName,
            "yl:invisible": isLoading
          })}
        >
          {!isLoading && icon?.iconName && icon.iconPosition === "left" && (
            <Icon
              className={classNames(
                {
                  "yl:w-5": !icon?.iconClassName
                },
                icon?.iconClassName
              )}
              iconName={icon.iconName}
            />
          )}
          {children}
          {!isLoading && icon?.iconName && icon.iconPosition === "right" && (
            <Icon
              className={classNames(
                {
                  "yl:w-5": !icon?.iconClassName
                },
                icon?.iconClassName
              )}
              iconName={icon.iconName}
            />
          )}
        </span>
        <span className='yl:absolute'>
          {isLoading && <Spinner className={classNames("yl:h-8 yl:w-8")} />}
        </span>
      </div>
    </button>
  );
};

export default Button;
