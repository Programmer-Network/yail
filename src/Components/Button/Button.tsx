import classNames from "classnames";

import Icon from "../Icon";
import Spinner from "../Spinner";
import { getButtonClasses } from "./Button.utils";
import { ButtonVariantEnum, IButtonProps } from "./types";

const Button: React.FC<IButtonProps> = (
  {
    onClick,
    children,
    className,
    disabled = false,
    type = "button",
    icon,
    isLoading = false,
    outlined = false,
    variant = ButtonVariantEnum.PRIMARY
  } = {
    onClick: () => null,
    children: null,
    className: "",
    disabled: false,
    outlined: false,
    isLoading: false,
    variant: ButtonVariantEnum.PRIMARY
  }
) => {
  return (
    <button
      disabled={disabled}
      onClick={onClick}
      type={type}
      className={getButtonClasses({
        variant,
        disabled,
        outlined,
        className,
        isLoading
      })}
    >
      <div className='relative flex items-center justify-center'>
        <span
          className={classNames({
            "flex items-center gap-1": icon?.iconName,
            invisible: isLoading
          })}
        >
          {!isLoading && icon?.iconName && icon.iconPosition === "left" && (
            <Icon
              className={classNames(
                {
                  "w-5": !icon?.iconClassName
                },
                icon?.iconClassName
              )}
              iconName={icon.iconName}
            />
          )}
          {children}
          {!isLoading && icon?.iconName && icon.iconPosition === "right" && (
            <Icon
              className={classNames(
                {
                  "w-5": !icon?.iconClassName
                },
                icon?.iconClassName
              )}
              iconName={icon.iconName}
            />
          )}
        </span>
        <span className='absolute'>
          {isLoading && <Spinner className={classNames("h-8 w-8")} />}
        </span>
      </div>
    </button>
  );
};

export default Button;
