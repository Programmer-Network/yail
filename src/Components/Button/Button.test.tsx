import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import Button from ".";

describe("Button component", () => {
  test("renders correctly - snapshot test", () => {
    const { asFragment } = render(<Button>Snapshot Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders button text", () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText("Click me")).toBeInTheDocument();
  });

  test("is disabled when disabled prop is true", () => {
    render(<Button disabled>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toBeDisabled();
  });

  test("shows spinner when isLoading is true", () => {
    render(<Button isLoading={true}>Click me</Button>);
    expect(screen.getByTestId("icon-spinner")).toBeInTheDocument();
  });

  test("has outlined styles when outlined prop is true", () => {
    render(<Button outlined>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-border-2");
    expect(button).toHaveClass("yl-text-primary");
    expect(button).toHaveClass("yl-border-primary");
  });

  test("handles click events", () => {
    const handleClick = vi.fn();
    render(<Button onClick={handleClick}>Click me</Button>);
    fireEvent.click(screen.getByText("Click me"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test("has correct styles when disabled and filled", () => {
    render(<Button disabled>Disabled Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-text-background");
    expect(button).toHaveClass("yl-bg-primary");
  });

  test("has correct styles when loading and filled", () => {
    render(<Button isLoading>Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("hover:yl-bg-transparent");
  });

  test("has correct styles when not disabled, not loading, and filled", () => {
    render(<Button>Button</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-bg-primary");
    expect(button).toHaveClass("hover:yl-bg-transparent");
    expect(button).toHaveClass("hover:yl-text-primary");
  });

  test("has correct styles when disabled and outlined", () => {
    render(
      <Button disabled outlined>
        Button
      </Button>
    );
    const button = screen.getByRole("button");
    expect(button).toHaveClass("yl-text-primary");
  });
});
