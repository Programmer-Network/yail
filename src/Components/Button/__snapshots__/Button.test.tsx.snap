// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > renders correctly - snapshot test 1`] = `
<DocumentFragment>
  <button
    class="yl:select-none yl:px-3 yl:py-2 yl:font-semibold yl:tracking-tight yl:rounded-md yl:border-2 yl:transition-all yl:duration-200 yl:transform yl:hover:scale-105 yl:cursor-pointer yl:border-primary yl:bg-primary yl:text-background yl:hover:bg-transparent yl:hover:text-primary"
    type="button"
  >
    <div
      class="yl:relative yl:flex yl:items-center yl:justify-center"
    >
      <span
        class=""
      >
        Snapshot Button
      </span>
      <span
        class="yl:absolute"
      />
    </div>
  </button>
</DocumentFragment>
`;
